synapse/handlers/profile.py

```diff
@@ -336,12 +336,18 @@ async def check_avatar_size_and_mime_type(self, mxc: str) -> bool:
         """Check that the size and content type of the avatar at the given MXC URI are
         within the configured limits.
 
+        If the given `mxc` is empty, no checks are performed. (Users are always able to
+        unset their avatar.)
+
         Args:
             mxc: The MXC URI at which the avatar can be found.
 
         Returns:
              A boolean indicating whether the file can be allowed to be set as an avatar.
         """
+        if mxc == "":
```

Comment:
I *think* that you might be able to get a non-string value into here via `RoomMemberHandler.update_membership_locked` -- does that seem correct or am I missing something?

Reply:
The avatar_url must have a `len`:

https://github.com/matrix-org/synapse/blob/8533c8b03d8916e3805c7d0e0020226017680147/synapse/handlers/room_member.py#L618

And given that it comes from JSON, that means that `avatar_url` is either a dict, list or string. 

But this isn't the first unverified body parameter. Frankly I'd just toss this one in the #8445 bucket.

Reply:
Ahh sorry, you're quite right (I didn't spot the `or ""`). So that's every JSON value other than integers and `true` then!

Reply:
I think the proper solution might be to do `avatar_url = content.get("avatar_url") or ""` and then use the value directly?

Reply:
I sympathise, but I'm mainly interested in fixing https://github.com/matrix-org/synapse/issues/12257 here. The worst that happens today when going via the RoomMember path is an internal server error when a client gives us a garbage `avatar_url`. I'd rather see a systematic approach to validation rather than an ad-hoc point fix.

A little more context: when trying to deactivate+erase a user, we set their avatar_url to `""` here:

https://github.com/matrix-org/synapse/blob/15382b1afad65366df13c3b9040b6fdfb1eccfca/synapse/handlers/deactivate_account.py#L149-L151

And that ends up failing if avatar restrictions are turned on.

Reply:
Yes, this should handle blank strings properly, but it seems silly to not also handle the other data while we're here.

Reply:
I'm not following you here---what other data you have in mind? Would you prefer `check_avatar_size_and_mime_type` to take `mxc: Optional[str]` instead of `mxc: str`? Or maybe `mxc: Any`?

Reply:
I'm suggesting add additional handling to raise errors if the value is not a string and not None, similar to what we do at https://github.com/matrix-org/synapse/blob/300ed0b8a6050b5187a2a524a82cf87baad3ca73/synapse/handlers/profile.py#L296-L304

Although that case seems to be rejecting strings... /me sighs. I'm not sure what the behavior of this code is supposed to be (should it accept nulls or not). Maybe we should just accept this as is.

Reply:
> I'm not sure what the behavior of this code is supposed to be (should it accept nulls or not).

Nor I; if it were up to me [we'd reject nulls here](https://github.com/matrix-org/synapse/issues/12257#issuecomment-1074260390).